source/fuzz/fuzzer_pass_split_blocks.cpp

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <utility>
+#include <vector>
+
+#include "source/fuzz/fuzzer_pass_split_blocks.h"
+#include "source/fuzz/transformation_split_block.h"
+
+namespace spvtools {
+namespace fuzz {
+
+using opt::IRContext;
+
+FuzzerPassSplitBlocks::FuzzerPassSplitBlocks(
+    opt::IRContext* ir_context, FactManager* fact_manager,
+    FuzzerContext* fuzzer_context,
+    protobufs::TransformationSequence* transformations)
+    : FuzzerPass(ir_context, fact_manager, fuzzer_context, transformations) {}
+
+FuzzerPassSplitBlocks::~FuzzerPassSplitBlocks() = default;
+
+void FuzzerPassSplitBlocks::Apply() {
+  // Collect up all the blocks in the module.
+  std::vector<opt::BasicBlock*> available_blocks;
+  for (auto& function : *GetIRContext()->module()) {
+    for (auto& block : function) {
+      available_blocks.push_back(&block);
+    }
+  }
+
+  while (!available_blocks.empty()) {
+    auto block_index = GetFuzzerContext()->GetRandomGenerator()->RandomUint32(
+        (uint32_t)available_blocks.size());
+    auto block = available_blocks[block_index];
+    available_blocks.erase(available_blocks.begin() + block_index);
+    if (GetFuzzerContext()->GetRandomGenerator()->RandomPercentage() >
+        GetFuzzerContext()->GetChanceOfSplittingBlock()) {
+      continue;
+    }
+    // We are going to try to split this block.  We now need to choose where to
+    // split it.  We do this by finding a base instruction that has a result id,
+    // and an offset from that base instruction.  We would like offsets to be as
+    // small as possible and ideally 0 - we only need offsets because not all
+    // instructions can be identified by a result id.
```

Comment:
Instruction have a unique id that can be obtained from the function `Instruction::unique_id()`.  Using it might simplify the code.

Reply:
This is ok if all the transformations are done on the same module, but in order to replay the transformations, this is probably not reliable so it depends on how the information gets used.

Reply:
It would indeed be great if we could use this, but for the reasons @alan-baker mentions I don't think it will work.  In order to facilitate shrinking fuzzed shaders by replaying subsets of transformations, it's important that transformations are as independent as possible.

For example, if we split blocks A and B, and having split block B are able to perform some other transformation T that triggers a shader compiler bug but that does not depend on having split A, we would like to be able to replay just the split of B and the further transformation T that it enabled.  But I believe that if we rely on Instruction::unique_id(), the fact that we no longer split A will mean that different unique ids would be used when splitting B, and then the ids that the later transformation T may depend on would not be there.

Reply:
So, correct me if I'm wrong but: I believe the Instruction::unique_id is generated at parse time. So, as long as the starting SPV is the same, the unique_id's setup at the start of the process should always be the same. 

If your starting SPV changes, then yes your unique_id changes, but then so would result_id + offset (like if a block was split on the starting SPV the offset would differ, yea?).